File: 品味Java的21種設計模式/dp_cpp/iterator/example5/ArrayIteratorImpl.cpp
#include "ArrayIteratorImpl.h"

namespace cn
{
	namespace javass
	{
		namespace dp
		{
			namespace iterator
			{
				namespace example5
				{

					ArrayIteratorImpl::ArrayIteratorImpl(SalaryManager *aggregate)
					{
						this->aggregate = aggregate;
					}

					bool ArrayIteratorImpl::hasNext()
					{
						//判断是否还有下一个元素
						if(aggregate!=0 && index<aggregate->size())
						{
							return true;
						}
						return false;
					}

					object *ArrayIteratorImpl::next()
					{
						object *retObj = 0;
						if(hasNext())
						{
							retObj = aggregate->get(index);
							//每取走一个值，就把已访问索引加1
							index++;
						}
						return retObj;
					}

					void ArrayIteratorImpl::remove()
					{
						//暂时可以不实现		
					}
				}
			}
		}
	}
}

File: 品味Java的21種設計模式/dp_java/iterator/example5/ArrayIteratorImpl.java
package cn.javass.dp.iterator.example5;

import java.util.Iterator;
/**
 * 用来实现访问数组的迭代接口
 */
public class ArrayIteratorImpl implements Iterator {
    /**
     * 用来存放被迭代的聚合对象
     */
    private SalaryManager aggregate = null;
    /**
     * 用来记录当前迭代到的位置索引
     */
    private int index = 0;

    public ArrayIteratorImpl(SalaryManager aggregate) {
        this.aggregate = aggregate;
    }


    public boolean hasNext() {
        //判断是否还有下一个元素
        if (aggregate != null && index < aggregate.size()) {
            return true;
        }

        return false;
    }


    public Object next() {
        Object retObj = null;

        if (hasNext()) {
            retObj = aggregate.get(index);
            //每取走一个值，就把已访问索引加1
            index++;
        }

        return retObj;
    }


    public void remove() {
        //暂时可以不实现
    }
}
